File: Common/ComponentInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;

namespace Common
{
    public class ComponentInfo : DistributableObject
    {
        #region Private Data Members
        // Define this, the Message class, identifier
        private static Int16 ClassId { get { return (Int16)DISTRIBUTABLE_CLASS_IDS.ComponentInfo; } }
        #endregion

        #region Public Properties and Other Stuff
        public enum PossibleAgentType { BrilliantStudent = 1, ExcuseGenerator = 2, WhiningSpinner = 3, ZombieProfessor = 4 };

        public Int16 Id { get; set; }
        public PossibleAgentType AgentType { get; set; }  
        public EndPoint CommmunicationEndPoint { get; set; }
        public StatusInfo Status { get; set; }
        public static int MinimumEncodingLength
        {
            get
            {
                return 4                // Object header
                       + 2              // Id
                       + 1              // Agent Types
                       + 1              // CommunicationEndPoint
                       + 1;             // Status
            }
        }
        #endregion
      
        #region Constructors
        public ComponentInfo() {}

        public ComponentInfo(Int16 id, PossibleAgentType type)
        {
            Id = id;
            AgentType = type;
        }

        /// <summary>
        /// Factor method to create an object of this class from a byte list
        /// </summary>
        /// <param name="bytes">A byte list from which the distributable object will be decoded</param>
        /// <returns>A new object of this class</returns>
        new public static ComponentInfo Create(ByteList bytes)
        {
            ComponentInfo result = new ComponentInfo();
            result.Decode(bytes);
            return result;
        }
        #endregion

        #region Encoding and Decoding methods

        /// <summary>
        /// This method encodes an object of this class into a byte list
        /// </summary>
        /// <param name="bytes"></param>
        public override void Encode(ByteList bytes)
        {
            bytes.Add(ClassId);                             // Write out the class type

            Int16 lengthPos = bytes.CurrentWritePosition;   // Get the current write position, so we
                                                            // can write the length here later

            bytes.Add((Int16) 0);                           // Write out a place holder for the length

            bytes.AddObjects( (byte) AgentType, Id);
            bytes.Add(CommmunicationEndPoint);
            bytes.Add(Status);

            Int16 length = Convert.ToInt16(bytes.CurrentWritePosition - lengthPos - 2);
            bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        
        }

        /// <summary>
        /// This method decodes a message from a byte list.  It can onlt be called from within the class hierarchy.
        /// </summary>
        /// <param name="messageBytes"></param>
        protected override void Decode(ByteList bytes)
        {
            if (bytes == null || bytes.RemainingToRead < MinimumEncodingLength)
                throw new ApplicationException("Invalid byte array");
            else if (bytes.PeekInt16() != ClassId)
                throw new ApplicationException("Invalid class id");
            else
            {
                Int16 objType = bytes.GetInt16();
                Int16 objLength = bytes.GetInt16();

                bytes.SetNewReadLimit(objLength);

                AgentType = (PossibleAgentType) bytes.GetByte();
                Id = bytes.GetInt16();
                CommmunicationEndPoint = bytes.GetDistributableObject() as EndPoint;
                Status = bytes.GetDistributableObject() as StatusInfo;

                bytes.RestorePreviosReadLimit();
            }
        }

        #endregion
    }
}

File: bsvszp/src/Common/ComponentInfo.java
package Common;

import java.net.UnknownHostException;
import org.omg.CORBA.portable.ApplicationException;

public class ComponentInfo extends DistributableObject
{
	private static short ClassId;
	private short Id;
    private PossibleAgentType AgentType;  
    private EndPoint CommmunicationEndPoint;
    private StatusInfo Status;
    private static int MinimumEncodingLength;
    
    public enum PossibleAgentType
    { 
    	BrilliantStudent(1),
    	ExcuseGenerator(2),
    	WhiningSpinner(3),
    	ZombieProfessor(4);
    	
    	private int value;
    	PossibleAgentType(int value)
    	{
    		this.value = value;
    	}
    	 public int getValue()
         {
         	return value;
         }
         
         public static short getStringValueFromInt(int i) {
             for (PossibleAgentType status : PossibleAgentType.values()) {
                 if (status.getValue() == i) {
                     return (short) status.value;
                 }
             }
             // throw an IllegalArgumentException or return null
             throw new IllegalArgumentException("the given number doesn't match any Status.");
         }
         public static PossibleAgentType convert(byte value) {
             return PossibleAgentType.values()[value];
         }    
         public static PossibleAgentType fromByte(byte b) 
         {
        	 PossibleAgentType temp = null;
             for(PossibleAgentType t : PossibleAgentType.values())
             {
                 if( t.value== (int)b)
                     temp = t;
             } 
             return temp;  //or throw exception
         }
         
         
        
    }

    public ComponentInfo() {}

    public ComponentInfo(short id, PossibleAgentType type)
    {
        setId(id);
        setAgentType(type);
        printAll();
    }

    private void printAll() {
    	getClassId();
    	getMinimumEncodingLength();
		
	}

	// new
    public static ComponentInfo Create(ByteList bytes) throws ApplicationException, Exception
    {
        ComponentInfo result = new ComponentInfo();
        result.Decode(bytes);
        return result;
    }
    
    @Override
    public void Encode(ByteList bytes) throws UnknownHostException, Exception
    {
        bytes.Add(ComponentInfo.getClassId());                             // Write out the class type

        short lengthPos = bytes.getCurrentWritePosition();   // Get the current write position, so we
                                                        // can write the length here later

        bytes.Add((short) 0);                           // Write out a place holder for the length

        bytes.AddObjects( (byte) AgentType.getValue(), getId());
        bytes.Add(CommmunicationEndPoint);
        bytes.Add(Status);
		
        short length = (short)(bytes.getCurrentWritePosition() - lengthPos - 2);
        bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        
    }
    
	@Override
	protected void Decode(ByteList bytes) throws ApplicationException, Exception
    {
    	if (bytes == null || bytes.getRemainingToRead() < ComponentInfo.getMinimumEncodingLength())
    		throw new ApplicationException("Invalid byte array", null);
		else if (bytes.PeekInt16() != getClassId())
				throw new ApplicationException("Invalid class id", null);
		else
		{
			short objType = bytes.GetInt16();
			short objLength= bytes.GetInt16();
				    	
			bytes.SetNewReadLimit(objLength);
				    
			setAgentType((PossibleAgentType.fromByte(bytes.GetByte())));
			setId((bytes.GetInt16()));
			setCommmunicationEndPoint(((EndPoint) bytes.GetDistributableObject())); // EndPoint.Create(bytes);
			setStatus((StatusInfo)bytes.GetDistributableObject()); //	StatusInfo.Create(bytes);

		    bytes.RestorePreviosReadLimit();
		}
    }
	
	public short getId() {
		return Id;
	}

	public void setId(short id) {
		Id = id;
		System.out.println("ComponentInfo.Id " + Id);
	}

	public PossibleAgentType getAgentType() {
		return AgentType;
	}

	public void setAgentType(PossibleAgentType agentType) {
		AgentType = agentType;
	}

	public EndPoint getCommmunicationEndPoint() {
		return CommmunicationEndPoint;
	}

	public void setCommmunicationEndPoint(EndPoint commmunicationEndPoint) {
		CommmunicationEndPoint = commmunicationEndPoint;
	}

	public StatusInfo getStatus() {
		return Status;
	}

	public void setStatus(StatusInfo status) {
		Status = status;
	}

	public static int getMinimumEncodingLength() 
	{
		
		MinimumEncodingLength =  4             			// Object header
                					+ 2           		  	// Id
                					+ 1          		  	// Agent Types
                					+ 1 					//EndPoint.MinimumEncodingLength
                					+ 1; 					//Tick.MinimumEncodingLength;
		System.out.println("ComponentInfo.MinimumEncodingLength =" + MinimumEncodingLength);
		return MinimumEncodingLength;
	}

	public static short getClassId() {
    	ClassId =  (short)DISTRIBUTABLE_CLASS_IDS.ComponentInfo.getValue();
    	System.out.println("ComponentInfo.ClassId " + ClassId);
    	return ClassId;
	}
}
